iptables: chains creation and deletion

azmeuk opened the pull request:
##### SUMMARY
Since #25099 in 2017 there is a demand for chain creation and deletion. A patch has been provided back then with #32158, but it was not tested, and the code cannot be merged in this state.

This patch is inspired by #32158, and allows iptables chain creation and deletion.
If the `state` parameter is `present` then the chain is automatically created if needed.
If the `state` is absent and no other parameter except `table` is passed, then the chain is deleted.

Please let me know if this behavior suits you, an alternative could be to add a boolean parameter that would trigger chain creation and deletion when true.

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
iptables

azmeuk committed:
iptables - added a ``chain_management`` parameter to control chain
creation and deletion

fixes #25099
closes #32158

azmeuk commented:
@sivel @Akasurde FYI

nitzmahone commented:
Thanks- the overall implementation looks good to us at a glance.

After some discussion, we're thinking that since the module has existed with the current behavior for so long, the potential for issues with the implicit behavior is worrisome... We'd probably be more comfortable with the explicit args to control this behavior. Also, would you feel comfortable writing integration tests to run against RHEL VMs?

bcoca commented:
Also a changelog is missing

bikeshedding new option name: `implicit_chain_management: false|true`
for when deleting/creating rules create/delete(empty) chain as needed

the example could work 'as is', just needs detection that there are no rules or other operations:
```
- iptables: chain=CUSTOMONE state=absent
``` 
the module should be able to create/delete the chain if no rules specified, we might want extra 'force' flag if the chain has any rules still and you want to delete it (rm -rf dirname/ ?)

azmeuk commented:
Thanks for reviewing. I will give a look to the integration tests.
@bcoca just to be sure I understand, can you confirm that what you are suggesting is:
- adding a `implicit_chain_management` boolean parameter that would defaults to `false`
- create or delete a chain when `implicit_chain_management` is `true`
- create or delete a chain when `implicit_chain_management` is `false` but no other parameter is passed

moon-bits commented:
@azmeuk: appreciate that PR! Thank you very much! Eagerly waiting for that feature.

Hopefully, it does not end like the PR before

azmeuk commented:
Gentle ping @nitzmahone and @bcoca :pray:
Can you please confirm that my understanding of your suggestions is good?

bcoca commented:
@azmeuk yes, sorry for delay, that seems to be what we want

azmeuk commented:
Finally I added a `chain_management` parameter, and chains are created and deleted only when `chain_management` is explicitly true.

I have written a basic integration test for iptables. There was no iptables integration test before, but I have just written a test for the `chain_management` parameter.

Please tell me if you want me to test more exotic situations.

Also, [some integration tests](https://dev.azure.com/ansible/ansible/_build/results?buildId=33639&view=logs&j=56a4f649-5f96-5169-2677-d4d17ca97982&t=ab982947-958a-5534-12f6-45aecb691644&l=536) are failing due to iptables not being installed. I am not sure what to do here, should I install iptables for all the packet manager for all the distros?

@nitzmahone @bcoca

bcoca commented:
as part of requirements, but for others you just need to skip the tests (Windows, OS X and BSDs ...)

azmeuk commented:
I finally could make the integration test work. I think this patch is ready now.
@bcoca @nitzmahone

azmeuk commented:
@bcoca gentle ping :pray:
